resources/resources.go

```diff
@@ -0,0 +1,380 @@
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+//An operator is interested in how many of a resource it manages and how many are ready
+type ResourceFigures struct {
+	Total int32
+	Ready int32
+}
+
+//In order to status check resources we check their status conditions for a condition
+//with a specific type and status. This type lets us define and pass those values around
+type ResourceConditionReadyRequirements struct {
+	Type   string
+	Status string
+}
+
+//Represents the metadata we pull off the unstructured resource
+type ResourceMetadata struct {
+	Generation      int64
+	Namespace       string
+	Name            string
+	UID             string
+	ResourceVersion string
+	OwnerReferences []string
+}
+
+//Represents the status we pull off the unstructured resource
+//I only parse what I need to reduce the oppurtunity for bugs
+type ResourceStatus struct {
+	ObservedGeneration int64
+}
+
+//Represents a k8s resource in a type-neutral way
+//We used to have lots of repeated code because we needed to perform
+//the same operations on different resources, which are represented by
+//the go k8s sdk as different types. This type provides a single
+//type that can represent any kind of resource and perform common
+//actions
+type Resource struct {
+	Source         unstructured.Unstructured
+	Status         ResourceStatus
+	Metadata       ResourceMetadata
+	Conditions     []map[string]string
+	ConditionClass ResourceConditionReadyRequirements
+}
+
+//Sets the resource ready requirements
+func (r *Resource) SetReadyRequirements(class ResourceConditionReadyRequirements) {
+	r.ConditionClass = class
+}
+
+//Returns true of the resource is owned by the given guid
+func (r *Resource) IsOwnedBy(ownerUID string) bool {
+	for _, ownerRef := range r.Metadata.OwnerReferences {
+		if ownerRef == ownerUID {
+			return true
+		}
+	}
+	return false
+}
+
+//Parses a resource unstructured object to populate this Resource object
+func (r *Resource) Parse(uObject unstructured.Unstructured) {
+	r.Source = uObject
+	r.parseMetadata()
+	r.parseStatusConditions()
+	r.parseStatus()
+}
+
+//Get the ready status
+func (r *Resource) IsReady() bool {
+	return r.readyConditionFound() && r.generationNumbersMatch()
+}
+
+//Returns true of the ready conditions are found
+//We only care to find one matching condition. Not all need to match to be Ready
+func (r *Resource) readyConditionFound() bool {
+	for _, condition := range r.Conditions {
+		if condition["type"] == r.ConditionClass.Type && condition["status"] == r.ConditionClass.Status {
+			return true
+		}
+	}
+	return false
+}
+
+//Returns true of the generation numbers are correct
+func (r *Resource) generationNumbersMatch() bool {
+	return r.Metadata.Generation <= r.Status.ObservedGeneration
+}
+
+//Parses the unstructured source metadata into this Resource object's metadata map
+func (r *Resource) parseMetadata() {
+	rawMetadata := r.Source.Object["metadata"].(map[string]interface{})
```

Comment:
We might need to talk about this - I think this might just be a lack of knowledge of the k8s SDK on my part. I don't know what a metadata object is or how to convert one of these into one. I'll do some research and see if I can self solve but I might need some help. I'll hit you back before end of your day.

Reply:
Did one better - turns out unstructured has a bunch of nice methods for pulling data off in a nice safe way:

```go
	r.Metadata = ResourceMetadata{
		Generation:      r.Source.GetGeneration(),
		Namespace:       r.Source.GetNamespace(),
		Name:            r.Source.GetName(),
		UID:             string(r.Source.GetUID()),
		ResourceVersion: r.Source.GetResourceVersion(),
		OwnerUIDs:       ownerUIDs,
	}
```